### ISSUE TITLE
Erroneous checking of `@CreatesObligation` on anonymous class

### ISSUE THREAD

**Issue by msridhar** [kelloggm/object-construction-checker#368]
Test case:
```java
class COAnonymousClass {
  static class Foo {

    @CreatesObligation("this")
    void resetFoo() { }

    void other() {

      Runnable r = new Runnable() {
        @Override
        @CreatesObligation
        public void run() {
          resetFoo();
        }
      };
    }
  }
}
```
I checked this in on the `co-anonymous-class` branch.  The issue is that the code should not type check.  The call to `resetFoo()` in `run()` creates an obligation on `Foo.this`, not `this`.  I am not 100% sure what error we should report here.  We could run into the same issue with a non-static inner class or a lambda.

**Comment by kelloggm** [comment]
One could argue that this code is actually fine as-is, since the `Runnable` is never started. If we add a call to `run`, then I think we are actually going to get the outcome we want: we would force the programmer to write an `@CreatesObligation` annotation on `other()`.

In the more general case, I agree that we need to think about this more. Especially for lambdas, which I suspect we might be getting wrong, and for more complex versions of the same kind of thing that we have here - confusion between the various "this" objects in scope at different nesting levels.

**Comment by msridhar** [comment]
Is this version more directly problematic?

```java
class COAnonymousClass {
  static class Foo {

    @CreatesObligation("this")
    void resetFoo() { }

    void other() {

      Runnable r = new Runnable() {
        @Override
        @CreatesObligation
        public void run() {
          resetFoo();
        }
      };
      other2(r);
    }

    void other2(Runnable r) { r.run(); }
  }
}
```

It still passes the checker (I updated the test case on the branch).

**Comment by kelloggm** [comment]
> Is this version more directly problematic?

Yes, I think this example makes the unsoundness more clear.

**Comment by msridhar** [comment]
@kelloggm should we close this now?

**Comment by kelloggm** [comment]
Yes, this should have been closed when #368 was merged.
